tensorflow_probability/python/experimental/mcmc/rhat_reducer.py

```diff
@@ -0,0 +1,203 @@
+  can be inferred as shape `[2, 3, 4]` samples drawn from 5 independent Markov
+  chains. If chunking is enabled and `axis=1` then the same chain state can be
+  inferred as 3 samples of shape `[2, 4]` from 5 independent Markov chains. Note
+  how the chunking dimension indexes into only the event dimensions, and not the
+  batch dimension.
+
+  As with all reducers, RhatReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+
+  `RhatReducer` is meant to fit into the larger Streaming MCMC framework.
+  `RunningPotentialScaleReduction` in `tfp.experimental.stats` is
+  better suited for more generic streaming covariance needs. More precise
+  algorithmic details can also be found by referencing
+  `RunningPotentialScaleReduction`.
+  """
+
+  def __init__(self, num_parallel_chains, name=None):
+    """Instantiates this object.
+
+    Args:
+      num_parallel_chains: Integer number of independent Markov chains
+        ran in parallel.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., 'rhat_reducer').
+
+    Raises:
+      ValueError: if `num_parallel_chains < 2`. This results in undefined
+        intermediate variance calculations.
+    """
+    if num_parallel_chains < 2:
+      raise ValueError('At least 2 parallel chains are required for '
+                       'well-defined between_chain variances')
+    self._parameters = dict(
+        num_parallel_chains=num_parallel_chains,
+        name=name or 'rhat_reducer'
+    )
+
+  def initialize(self, initial_chain_state, initial_kernel_results=None):
+    """Initializes a `RhatReducerState` using previously defined metadata.
+
+    For calculation purposes, the `initial_chain_state` does not count as a
+    sample. This is a deliberate decision that ensures consistency across
+    sampling procedures (i.e. `tfp.mcmc.sample_chain` follows similar
+    semantics).
+
+    Args:
+      initial_chain_state: A (possibly nested) structure of `Tensor`s or Python
+        `list`s of `Tensor`s representing the current state(s) of the Markov
+        chain(s). It is used to infer the shape and dtype of future samples.
+      initial_kernel_results: A (possibly nested) structure of `Tensor`s
+        representing internal calculations made in a related `TransitionKernel`.
+        For streaming rhat, this argument has no influence on the
+        computation; hence, it is `None` by default. However, it's
+        still accepted to fit the `Reducer` base class.
+
+    Returns:
+      state: `RhatReducerState` with `rhat_state` field representing
+        a stream of no inputs.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'rhat_reducer', 'initialize')):
+      initial_chain_state = tf.convert_to_tensor(initial_chain_state)
```

Comment:
This is not consistent with the claim that the chain state can be an arbitrary structure of Tensors.  `tfp.mcmc.potential_scale_reduction_factor` seems to support at least a list of Tensors, which suggests that an arbitrary structure should not be difficult.  Can we do that?